generic: nvmem: replace u-boot-env endianness workaround

mcprat opened the pull request:
ping @musashino205 @rmilecki @Ansuel 

The previously proposed fix was to add further endianness conversion
for big-endian systems in order to match the expected crc32 value.
However, this worked because the expected value had endianness conversion
when it was actually not necessary, and should be removed instead.

Ref: 797177ad8 ("kernel: update U-Boot nvmem driver to v6.2 release version")
Link: https://git.kernel.org/torvalds/linux/c/2d7521aa26ec

mcprat committed:
kernel: nvmem: replace u-boot-env endianness workaround with accepted fix

The previously proposed fix was to add further endianness conversion
for big-endian systems in order to match the expected crc32 value.
However, this worked because the expected value had endianness conversion
when it was actually not necessary, and should be removed instead.

Ref: 797177ad8 ("kernel: update U-Boot nvmem driver to v6.2 release version")
Link: https://git.kernel.org/torvalds/linux/c/2d7521aa26ec
Signed-off-by: Michael Pratt <mcpratt@pm.me>

musashino205 commented:
LGTM.

leobsky commented:
@mcprat with the upcoming kernel 6.6.100 nvmem-u-boot-env-remove-crc32-endianness-conversion.patch  was moved to `/drivers/nvmem/u-boot-env.c`
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/commit/drivers/nvmem?h=linux-6.6.y&id=7792a10f80a3ce4ecf57036988599bc9b4d9b680

mcprat commented:
@leobsky This is true for upstream backports, however, we have even more backports than upstream which moves it to a layout driver just like linux main tree.

mcprat commented:
@hauke please merge to undo this mess

a proper bump to 6.6 will either:

1. manually refresh only the removal lines of the "add layout driver" patch and preserve my patch, or,
2. add a revert patch of the upstream backport so neither of the other patches need to be modified.

a proper bump to 6.12 will remove my patch as a backport instead of the old one...

leobsky commented:
@mcprat perhaps is better leave 6.12 as it now, patch is already upstreamed.
6.6 only need the revert patch https://github.com/leobsky/openwrt/pull/1/commits

mcprat commented:
except that you're messing up the git history and the bump for both kernels look different without any reason, and both of them should be nearly identical and different than they are now

robimarko commented:
Please drop the kernel update revert

mcprat commented:
it's there because last kernel bump is messing up the git history, mashing things that should be separate patches together, and commit message not accurate as a result

KanjiMonster commented:
The endian fix was included in [6.6.100](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.6.100&id=48e8791843206baf76827df1b6ee3cb88a2a17d8) and [6.12.40](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.12.40&id=c29a2328af96338d327cd851803338423c6f07a1), so with these bumps there is no need for any backported patches anymore. Bumping the kernel is essentially equivalent with replacing it with the accepted version.

A re-applied bump to 6.6.100 and 6.12.40 would just result in the same state as before, and I fail to see anything wrong with them.

KanjiMonster commented:
So then just fix up the patches (or probably just delete the old versions that were left?). I really don't see the need for a full revert of the kernel bumps.

mcprat commented:
really there's nothing in git history currently that explains why the patch is removed because it doesn't exist upstream

mcprat commented:
by the way, letting any patch at any time be removed from "pending" because it was backported without being added to "backports" in the same commit is a bad habit that should stop and would prevent this

KanjiMonster commented:
> by the way, letting any patch at any time be removed from "pending" because it was backported without being added to "backports" in the same commit is a bad habit that should stop and would prevent this

That would require actively tracking all pending patches and most times noise, unless the accepted version is (significantly) different from the pending one.

But if a kernel bump includes a patch that is in pending/ then there is no way how that patch could exist a swell in `backports` - backports is for patches that didn't get backported to the kernel version (yet).

mcprat commented:
I'll do that part even, give me a few hours

KanjiMonster commented:
> no it's easy, because while working on a kernel bump the conflict with "patch can be reverse applied" would appear, telling you which patch is a backport, then one does the small research to see what happened and move it to backports before doing the bump

I really don't think this is worth the extra commit(s) and work for contributers. And especially not worth a revert.

> anyway, I believe the final fix should be in our git history so it can be understood what happened without digging into the kernel, and there's no way to do that without a revert

And I really don't see the point of it. Nobody will use that commit (unless accidentally while bisecting), so the "final" version will never be built or tested or anything since it will be immediately deleted.

And even if I were to agree to you with having the commits, I wouldn't accept this without there being a re-application of the kernel bumps so that the kernel version at the end is again up to date.

mcprat committed:
Revert "kernel: bump 6.12 to 6.12.40"

This reverts commit b6276e33eb02f13ea955cdf1c927b71ace1edf05.

Instead of transitioning to a new patch and then removing it later
as a backport, an obsolete patch is quoted as being accepted.
Redo the bump to preserve git history.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
Revert "kernel: bump 6.6 to 6.6.100"

This reverts commit 05d344cb2ca3b95f6c3536981af4fd85122eae0d.

A fix for nvmem was not properly handled by transitioning
to a backport patch before doing this bump.
Not removing the endianness nvmem patch here results in
nvmem being broken for u-boot environment parsing.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
kernel: bump 6.6 to 6.6.100

Manually refreshed to undo a manual backport:
 - 823-v6.12-0003-nvmem-layouts-add-U-Boot-env-layout.patch

All other patches automatically refreshed.

Link: https://git.kernel.org/stable/linux/c/48e879184320
Link: https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.6.100
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
kernel: move accepted net offload patch to backports

This patch is accepted without changes
and backported into the next 6.12 release.

Copy from upstream and automatically refresh.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
kernel: bump 6.12 to 6.12.40

Upstream backports removed:
 - 680-v6.16-net-fix-TCP-UDP-fraglist-GRO.patch
 - 825-v6.16-nvmem-u-boot-env-remove-crc32-endianness-conversion.patch

All patches automatically refreshed.

Link: https://git.kernel.org/torvalds/linux/c/9f735b6f8a77
Link: https://git.kernel.org/torvalds/linux/c/2d7521aa26ec
Link: https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.12.40
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat commented:
ok how does this look @KanjiMonster @hauke 

I did not mean to impose extra work onto other people, I just want a consistent git history for file blaming and explanations in commit messages that make sense.

note that the actual backport patch stays in 6.6 now while the old proposed patch is removed.

mcprat commented:
one moment, something I tried with the script caused some to not refresh

mcprat commented:
@hauke can I at least get your thoughts on what should be done about the last bump?

if you merge this now it will resolve the issue without leaving the rest of the patches in a reversed state, as suggested.

sppmasterspp commented:
> For some reason, this community is suddenly afraid of a revert patch even though we have several, but perfectly fine with merging the bad work to begin with.

Maybe it's just a vacation time for the developers...

KanjiMonster commented:
AFAIU the core issue here was that the version in our tree and the upstream accepted version were different enough that the usual detection of "can be reverse-applied" did not trigger, as the old version had a different approach and modified different parts, and not breaking the build.

This resulted in the old version kept in our tree, essentially fixing the issue again, so breaking it by undoing the upstreamed fix.

I still think reverting is overkill, and a commit just dropping the obsolete/duplicated fixes with an appropriate commit message is enough.

And requiring people bumping kernels to go through all patches in pending (of which there are > 100) is not a reasonable request. We don't even provide the means of efficiently doing so.

If we ever want to require/encourage something like this, we need to make this easy, ideally with a script. E.g. add an upstream submission / patchwork link to the patch, then have a script that can go through all patches and check them for being accepted / rejected / superseded by a newer version, maybe even replacing them directly.
